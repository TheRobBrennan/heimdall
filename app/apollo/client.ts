import fetch from "cross-fetch"
import { useMemo } from "react"

import { ApolloClient, createHttpLink, InMemoryCache } from "@apollo/client"
import { setContext } from "@apollo/client/link/context"

let apolloClient

const httpLink = createHttpLink({
  uri: `api/graphql`,
  fetch,
})

/* istanbul ignore next */
const authLink = setContext(async () => {
  // See https://www.apollographql.com/docs/react/networking/authentication/#header for more details
  // Get our session detail from the back-end API
  let authorization = ""
  try {
    const data = await (await fetch("api/session")).json()
    const { accessToken } = data
    authorization = `Bearer ${accessToken}`
  } catch (e) {
    // No session data available; no authorization header will be sent
  }

  // Return the headers to the context so httpLink can read them
  return {
    headers: {
      authorization,
    },
  }
})

function createApolloClient(initialState) {
  const ssrMode = typeof window === "undefined"

  return new ApolloClient({
    ssrMode,
    link: authLink.concat(httpLink),
    cache: new InMemoryCache().restore(initialState),
  })
}

export function initializeApollo(initialState = null) {
  const _apolloClient = apolloClient ?? createApolloClient(initialState)

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // get hydrated here
  if (initialState) {
    _apolloClient.cache.restore(initialState)
  }
  // For SSG and SSR always create a new Apollo Client
  /* istanbul ignore next */
  if (typeof window === "undefined") return _apolloClient
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient

  return _apolloClient
}

/* istanbul ignore next */
export function useApollo(initialState) {
  const store = useMemo(() => initializeApollo(initialState), [initialState])
  return store
}
