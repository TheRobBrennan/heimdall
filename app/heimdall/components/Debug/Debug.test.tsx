import React from "react"
import TestRenderer from "react-test-renderer"
import { MockedProvider } from "@apollo/client/testing"

import Component from "./Debug"

import type { Auth0User } from "../../../auth0/user"

describe("Our Debug component", () => {
  describe("should render", () => {
    it("Auth0 user details", () => {
      const user: Auth0User = {
        given_name: "Rob",
        family_name: "Brennan",
        nickname: "rob",
        name: "Rob Brennan",
        picture:
          "https://lh3.googleusercontent.com/a-/AOh14GijQDyp8j2v_8Dg3WMs7BtVIqs-dDVizUnEuNLbPKI=s96-c",
        gender: "male",
        locale: "en",
        updated_at: "2021-01-24T04:46:38.864Z",
        sub: "google-oauth2|116058668302290861810",
      }

      const component = TestRenderer.create(
        <MockedProvider mocks={[]} addTypename={false}>
          <Component user={user} />
        </MockedProvider>
      )
      expect(component.toJSON()).toMatchSnapshot()
    })
  })
})
