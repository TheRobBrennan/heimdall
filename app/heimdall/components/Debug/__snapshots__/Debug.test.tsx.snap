// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Our Debug component should render Auth0 user details 1`] = `[Function]`;
